### ISSUE TITLE
z/VM on NUMA selectively turns CPs off when under moderate load

### ISSUE THREAD

**Issue by moshix** [SDL-Hercules-390/hyperion#130]
When I run z/VM 6.2 or 6.3 on Hercules with > 4 CPs and start two guest OSs at the same time, z/VM soon will start turning off some CPs (stopped state in hercules) and then crashes without dumping. 

I can reproduce this at will on a HP G580 with 4 populated sockets with Xeon E7-4870s (40 cores in total, 4 x 10, and 512GB of RAM).

**Comment by moshix** [comment]
Yesterday, all trouble seemed so far away (pre SDL-hypersion). Then they looked as though they were here to stay (when I started using SDL-hyperion on NUMA). How I long for yesterday. 

But then, I switched to Spinhawk 3.13 and poof all my NUMA trouble is gone!

**Comment by Fish-Git** [comment]
### _Breaking News!_
> I don't have NUMA hardware (that I am aware of!)

Apparently I _**do**_ have NUMA-capable hardware but just never realized it!!

I happened to stumble across some photographs of my system's BIOS settings that I took with my camera(*) and to my utter amazement saw this:

![dscn0764](https://user-images.githubusercontent.com/1486786/44896285-96411480-acac-11e8-83c7-e7b6140dfb99.jpg)
&nbsp;

I'm going to try enabling NUMA on my system(**) and see if maybe I can recreate some of the problems you're having.
___
_(*) I wish BIOS manufacturers would include some type of "Save BIOS settings report to file" function or something similar (like being able to save firware to a file, but generating a human readable text report instead in this particular case). It's a real PITA to have to reboot your system just to see what one of your BIOS settings happens to be set to!_

_(**) After doing a full system backup of course! I don't fully trust Windows with respect to it running in NUMA mode. I suspect 99.99% of Windows users running on hardware that just so happens to be NUMA-capable like mine, probably, like me, let it default to SMP mode instead. Thus I suspect Microsoft's NUMA logic is likely not very well tested. I'm going to have to be careful here. Especially after discovering the following!_

* **[KB 2155311](https://support.microsoft.com/en-us/help/2155311/poor-performance-occurs-on-a-computer-that-has-numa-based-processors-a): "Poor performance occurs on a computer that has NUMA-based processors and that is running Windows Server 2008 R2 or Windows 7 if a thread requests lots of memory that is within the first 4 GB of memory"**
&nbsp;

**Comment by moshix** [comment]
Ivan

Real NUMA is when you have several several sockets with memory in different locatiosn and only logically represented as linear. The above doesnt' seem to be a real NUMA. The NUMA system I use is a HP DL580 G9 which has 4 processor sockets and thus necessarily non-uniform, non-linear RAM modules organization. 

here is a link: https://www.cnet.com/products/hp-proliant-dl580-gen9-xeon-e7-8860v3-2-2-ghz-128-gb-0-gb/specs/

Windows Server generally has very poor implementation of NUMA support. Linux does a much better job. I run my DL580 with Red Hat Linux as the underlying OS, and then Debian guest machines. 

I can give you access to my DL580 if you would like, when i get back from my trip. Pls email me separately about it. 

thanks

**Comment by ivan-w** [comment]
Moshe,

That's because you are using vertical polarization (the z/VM default if the SET TOPOLOGY instruction supports it). When z/VM detects a light load, it will "park" processors that are not in use. If you find it problematic, issue a "SET SRM POLARIZATION HORIZONTAL".

z/VM putting unused processor is not an actual problem, it is an expected behavior on a vertical polarization system.

Vertical polarization under hercules isn't really "supported" on hercules (just emulated) since emulated CPU threads have no host CPU affinity (and thus cannot really take advantage of an underlying NUMA).

--Ivan

**Comment by moshix** [comment]
Ivan

wish it was so simple. It isn't, unfortunately. The mainframe machine becomes unsuable, unstable and ultimately unresponsive when that happens.  I will try to set polarization horizontal (the thought had occurred to me originally,but due to the crash, I discarded it). 

Also, given that Hercules 3.13 doesn't display this behavior, ever, I doubt this is a z/VM configuration or use problem. 

But I shall play with polarization to see if this somehow affects behavior to see if we are getting closer to the problem. 

thanks

M

**Comment by Peter-J-Jansen** [comment]
Without NUMA hardware I cannot reproduce the crash, but SET SRM POLARIZATION HORIZONTAL stops the following messages when running z/VM 6.3 : 
```
17:07:02 HHC00809I Processor CP05: disabled wait state 0102400080000000 0000000000000000
17:07:02 HHC00809I Processor CP04: disabled wait state 0102400080000000 0000000000000000
17:07:08 HHC00814I Processor CP00: SIGP Restart                          (06) CP04, PARM 000000000024E2F8: CC 0
17:07:30 HHC00814I Processor CP00: SIGP Restart                          (06) CP05, PARM 000000000024E2F8: CC 0
17:07:32 HHC00809I Processor CP04: disabled wait state 0102400080000000 0000000000000000
17:07:32 HHC00809I Processor CP05: disabled wait state 0102400080000000 0000000000000000
17:07:38 HHC00814I Processor CP00: SIGP Restart                          (06) CP04, PARM 000000000024E2F8: CC 0
17:07:48 HHC00814I Processor CP00: SIGP Restart                          (06) CP05, PARM 000000000024E2F8: CC 0
```
as Ivan described.

Cheers,

Peter

**Comment by moshix** [comment]
Peter

yes, I see the same, the messages disappear. But I cannot reproduce the bug at will, so I cannot exclude that the problem is still there, even with the horizontal polarization. Will stress it a bit more and see. 

The underlying issue with reproducing the bug is that it's not really predictive how Linux assigns RAM from outside the local front-end bus to an app. It depends on too many things, and so it's hard to reproduce the bug. 

thanks

**Comment by moshix** [comment]
ok, folks it happened again. 

Setup: real NUMA machine (HP DL580 gen9 with 512GB of RAM, 4 sockets with 6 cores each). Ubuntu 16.04. I start a Hercules instance with 8 CPs allocated to it, and therefore 2 CP will be on cores outside of a socket, therefore forcing NUMA memory access for 2 cores. 

I run zVM6.3 and start several zLinux guests (large ones, 6GB memory allocated to each VM). 

I do SET SRM POLARIZATION HORIZONTAL 

Under heavy load, 2 CPs went into STOPPED mode. the MIPS counter indicated zero MIPS, while Hercules was still furiously working on the remaining CPs. 

I went to grab my phone to take a video of this (like I did the first time), and when I got back Hercules had segfaulted. 

Tried immediately to reproduce the problem, and I couldn't force it to happen again. 

However, to me this feels a lot like a NUMA issue, specifically due to a race condition due to memory not responding fast enough. 

thanks

**Comment by Fish-Git** [comment]
I'm sorry you're still having problems, @moshix.&nbsp;`:(`

Trying to research this NUMA issue is on my TODO(tm) list, and I promise you I _will_ get "a round tuit" one of these days. Due to current workload however, it might still be a while.

Your patience in this matter is appreciated.

**Comment by Fish-Git** [comment]
> I have encountered such occurrences on non-NUMA system ... (no later than earlier today)

z/VM 6.3 (or some other version of z/VM) with SET SRM POLARIZATION HORIZONTAL active?

Interesting!

Does it otherwise work okay when you _don't_ specify horizontal polarization? Myabe that's where our bug is. Is that what you're saying, Ivan? That you believe there's a bug in our vertical/horizontal CPU polarization logic as related to z/VM (SIE) and NUMA or non-NUMA is just a red herring? (wold goose chase, false lead)? Is that what you're saying?

I want to be sure I understand you correctly!

**Comment by ivan-w** [comment]
Fish,

SET SRM POLARIZATION was VERTICAL at the time I encountered the problem. The CPUs may be going into a wait state for completely other reasons when the problem occurs.
Set Fact that Moshe's system is a NUMA system is possibly a red herring... It's more likely (wild guess) he is running into the problem more often than others because the beast he is using is a gigantic one !

I am going to set up something similar to what Moshe has (z/VM with multi linux guest running heavy workloads) and see if I can reproduce something similar.

There is also another problem and it MIGHT be related... The CPU Timer is still running amok. On a 1st level z/VM system, there are minor discrepancies, but when you run z/VM under z/VM, the virtual machines have a completely (and I mean *COMPLETELY*) CPU Timer erratic behavior.

Note : When I run into the problem described, performing a system reset, ipl clear and whatever does not fix the problem... hercules becomes completely unusable and has to be quit and restarted (the internals of hercules ARE broken at that point).

--Ivan

**Comment by moshix** [comment]
Ok, made some progress with this issue, folks. 

It turns out it's a memory constraint issue. Let me give you an example:

if I I have a Linux machine with 32GB of RAM, and 8 CPUs, and about 12 DASDs, then if I give Hercules 8 CPUS and 30GB of memory, I start to see the turning off of CPs in zVM 6.3

If instead I give in the above configuration, only 24GB of memory to hercules, things work fine, with polarization horizontal of verticial(!)

I made a quick calculation, the Linux kernel needs about 3GB for the page tables on a 32GB RAM system, then Hercules needs a bit of memory, too, and for every opened DASD, the LInux VFS file system abstraction laywer needs a few hunred MBs as well, quickly after IPL. 

So, I took the MAINSIZE gradually down until I never had stopped CPs anymore. On a 32GB machine, that was at MAINSIZE 24GB which has now been up and running under stress load, for about 2 weeks. 

So, consider this issue fixed. The resolution is to leave adequate RAM for Linux, and once paging starts, bad things happen to z/VM 6.3

z/VM 6.2 is much, much less sensitive to this issue, by the way. 

i can run with 29GB MAINSIZE on a 32GB machine and have no problems at all. 

thanks

**Comment by Fish-Git** [comment]
That's great news, Moshe!

 I'm glad you finally figured it out.

(I'm also glad it's not a Hercules issue too!)&nbsp;&nbsp;`;-)`

**Comment by moshix** [comment]
Fish

I did a lot of testing this week, and here is guide for what works and what doesnt' when it comes to memory allocation to Hercules. maybe we can include that in the documentation or on the Hyperion website. 

there are numbers which reliably keep zVM 6.3 and 6.2 up and running. Anything above that will at some point cause the selective turning off of CPUs (a sure sign that z/VM sees too much MAINSIZE and Linux starts paging)

For Linux kernel 4.40 and newer:

Hercules MAINSIZE= RAM of Linux machine * 0.85) - 2GB - (150MB * 3390 device)
example: Linux with 16GB RAM and 10 3390s: (16GB*0.85)-2GB - (150MB * 10) = MAINSIZE 10100 (in MB)

for kernel 3.xx (like Ubuntu 14.04), the multiplier is 0.90, and the system ram deductibel is 1.5GB, so 
Hercules MAINSIZE= RAM of Linux machine * 0.90) - 1.5GB - (150MB * 3390 device)

I haven't done any Windows testing. 

thanks

**Comment by moshix** [comment]
oh good! Yes it is an interrupt loop which causes CPs to be turned off. very good, Ivan!

solving this would be amazing. 

thanks

**Comment by ivan-w** [comment]
Getting very close !
The issue occurs because of a series of circumstances and a very subtle bug in Real to Absolute address conversion (APPLY_PREFIXING) under SIE (which leads to z/VM home address space Segment Table to become garbled when a SIE guest stores into a real address (after DAT) which closely looks like the host prefix address !) - And because when you have more than 2G, z/VM uses address 0x80000000 to store its own segment table ! 

So basically if the host prefix is X'132000' and a SIE guest tries to store at host real address X'180132000' (after guest and host DAT is applied), it will actually store data at X'80000000' (right in the middle of z/VM's Segment table !)

Now, all I have to do is find a way to fix this (but there are some subtle casting involved)...

--Ivan

**Comment by ivan-w** [comment]
Actually it's more obvious than that ! When an instruction is executing under SIE in S/390 mode, the prefix is applied to the host in S/390 mode as well, even if the guest is actually in z/Arch mode (meaning some portions of the address are masked out (bit 0-32 in the real address) and APPLY_PREFIXING will wrongly believe prefixing needs to be applied).

**Comment by Fish-Git** [comment]
Fish wrote:
> (I'm also glad it's not a Hercules issue too!)

Famous last words, apparently! (given Ivan's most recent comments!)

Great effort, Ivan!  THANK YOU!

 It sounds like you are on the right track and are very close to nailing this issue!

**Comment by ivan-w** [comment]
Fixed by 10fbb0dc9b57fbd40a5ebadaf2aa8741f577c1f9

**Comment by moshix** [comment]
Ivan

thanks a lot, and good thinking. 

However, after getting latest SDL hyperion and building it, I still get the same behavior (selectively turning off CPs), and I can reliably reproduce it on my laptop this way:

1. IPL zVM 5.3 with 2GB of RAM in Hercules
2. IPL guest with memory over-allocation (say zOS with 3GB for guest machine)
3. guest starts IPL process and then zVM 5.3 crashes with HTT001

There is enough paging space on the paging devices. 

I haven't tried 6.2 or 6.3 yet, because I am traveling, but will do so later next week. 

thanks

M

**Comment by Fish-Git** [comment]
Moshe:

> 1. IPL zVM 5.3 with 2GB of RAM in Hercules
> 2. IPL guest with memory over-allocation (say zOS with 3GB for guest machine)
> 3. guest starts IPL process and then zVM 5.3 crashes with HTT001

I noticed the same behavior myself when I originally started playing with z/VM 5.3 when IBM first released it back in April 2013:

> p.s. There appears to be a bug in either z/VM 5.3 or else Hercules that prevents the OSA from working properly with z/VM unless z/VM is given about 2500MB-2600MB or more of real storage (i.e. MAINSIZE). Likely related to this, the z/OS guest of z/VM fails to IPL properly if given more than about 2000MB of storage. I have played with each (increasing/decreasing both, increasing or decreasing just one but not the other) but the only setting that seems to work fairly reliably is 2000MB for both z/VM and the z/OS guest. But then z/VM's OSA doesn't work. So z/VM was given 2600MB so its OSA can then work. I have not been able to get the z/OS guest to work correctly with the OSA yet however. I think I might not have things (zVM config and/or zOS guest) configured properly. This is all new to me and I'm stumbling.

&nbsp;
I think it might be a bug in z/VM 5.3.

z/VM 6.3 seems to work fine using an OSA in Layer 2 or Layer 3 even with a MAINSIZE of 5G or more, although I have _not_ had time yet to try IPLing a z/OS guest under it.

**Comment by ivan-w** [comment]
Right now running a 16GB z/VM 6.3 instance .. with 640 virtual machines running a sieve program for the first 50,000,000 primes... and no abend or abnormal behavior in the past 2 hours !

**Comment by Fish-Git** [comment]
> The patch created by Ivan seems indeed to have fixed the issues I originally reported in this thread and others.

That's great news!  Thanks for letting us know and the effort involved in testing. Mcuh appreciated!

> The ghost MIPS indicator also seems fixed.

With my patch in place? Or without it?

**Comment by Fish-Git** [comment]
> Right now running a 16GB z/VM 6.3 instance ... with 640 virtual machines running a sieve program for the first 50,000,000 primes... and no abend or abnormal behavior in the past 2 hours!

_**HOLY CRAP, Ivan!**_ &nbsp; _640 instances?!_ &nbsp; That's insane! &nbsp;(but _great_ news!)

**Comment by moshix** [comment]
small update:

i am very happy with the stability now. I have some *very large guests*  of 48GB running and I force z/VM to page ( up to 200 pages per second) and none of the problems we had before. 

Well done Ivan!
